### ISSUE TITLE
Loading the font with @fontface?

### ISSUE THREAD

**Issue by Aetherpoint** [opentypejs/opentype.js#62]
Curious – is there anyway to access the font object with a CSS @fontface? Or would that require recompiling the font somehow?

**Comment by interactionist** [comment]
@fpirsch the 33% more additional base64 bytes should be mitigated by gzipping when you send the data ==> http://davidbcalhoun.com/2011/when-to-base64-encode-images-and-when-not-to/

**Comment by Pomax** [comment]
Note that if opentype.js is already running in-browser, then bear in mind that the data:uri will basically be immediately unpacked again in-memory, so you only save a bit of space in the `data:` string. The opentype.js font representation, and the font as loaded from `data:` source will both stay exactly the same sizes. For transmission, using gzipped tables inside the WOFF is a good idea. If it's all in the same client session, there's virtually no benefit.

**Comment by interactionist** [comment]
Ah yes, if it is all in the client that would make no sense. 
I thought they want to persist it in a CSS/HTML and transfer the font embedded over network.

**Comment by cyberwombat** [comment]
Has any of this been implemented?

**Comment by fdb** [comment]
OpenType.js doesn't have this yet, but Plumin.js (which is built on top of OpenType.js) uses [this technique](https://github.com/byte-foundry/plumin.js/blob/d0473591230fe7854e589aae14a199d8af86ea07/src/Font.js#L286).

**Comment by cyberwombat** [comment]
I see thank you. Is it possible to extract a base64 rendition of the font from OT once it's loaded? I may be able to use it that way instead

**Comment by cyberwombat** [comment]
Thanks. I was able to use the `toArrayBuffer()` and convert to base64 and load it that way in my css.

**Comment by fdb** [comment]
Yeah, we should probably add that to OpenType.js directly....

**Comment by Pomax** [comment]
As a small aside, given that this issue is grounded in the past (filed august 2014), when the webfont landscape was very different: these days you *do* want just WOFF (possibly with WOFF2), as `eot` is obsolete (MS no longer supports any browser that *needs* the `eot` format), `svg` has been disavowed (most browsers that still had support for it in 2014 have since removed support again), and `ttf`/`otf` are full system fonts, subject to potentially stricter validation than "explicitly for web" fonts.

**Comment by Connum** [comment]
I agree there should be an easy integrated way to use the (possibly modified) fonts for text preview in browser context. Doesn't have to be in 2.0.0, but it shouldn't be too complicated to have this soon in 2.x.x at least.

**Comment by ILOVEPIE** [comment]
@Connum there's some problems with this, first off I don't think it should be the responsibility of our library to handle this, as it would introduce side effects. Secondly, this is already possible using Object URLs on the generated TTF or OTF files and embedding a new stylesheet that references the object URL. One thing we could do however is support outputting WOFF fonts, that could be useful.

**Comment by Pomax** [comment]
WOFF2 at least, then. WOFF was superceded *quite* a while ago =)

**Comment by ILOVEPIE** [comment]
@Pomax WOFF2 requires some fancy compression that is out of scope for this library, there are libraries that can be used with this one to read WOFF2 files.

**Comment by Connum** [comment]
Currently, we only have download() as a public method. If you want to only get a URL, you have to use all the more low-level methods yourself, building the DataView from the ArrayBuffer, the Blob from the DataView and finally the ObjectURL from the blob (or build a data: URL yourself by base64-encoding the data). I think it wouldn't hurt to expose a helper method (that we could then reuse in the download() method to download the file). After all, the download() method could be regarded as out-of-scope as well, as you could to that yourself from the underlying data.

As all modern browsers support TTF/OTF directly, I would also see this decoupled from WOFF writing support, as it would be enough to simply preview the current state of the font object in an input/textarea/contenteditable element.

On a side node, we should probably call `revokeObjectURL()` in the download method to free up memory.

**Comment by Pomax** [comment]
@ILOVEPIE certainly (I wish JS had a Compression object, given that the browser already has in/deflate, gzip, and brotli built in), but if you're not going to support WOFF2, there's probably not a lot of benefit to supporting WOFF, as the older, superseded standard.

**Comment by ILOVEPIE** [comment]
> Currently, we only have download() as a public method. If you want to only get a URL, you have to use all the more low-level methods yourself, building the DataView from the ArrayBuffer, the Blob from the DataView and finally the ObjectURL from the blob

iirc can't you just create a blob from an arraybuffer by passing it into the blob constructor?

**Comment by ILOVEPIE** [comment]
@Pomax here's how you can export/import WOFF2 from opentype.js:
https://github.com/opentypejs/opentype.js/issues/183#issuecomment-1147228025

**Comment by Connum** [comment]
I just did a quick proof-of-concept here:
https://github.com/opentypejs/opentype.js/tree/font-face (_font-inspector.html_)

It's as easy as
```html
<style>
textarea {
  font-family: "OpentypeJsPreview";
}
</style>
<style id="opentypejs-preview-fontface"></style>
<textarea></textarea>
```
```js
document.getElementById('opentypejs-preview-fontface').innerHTML = `
@font-face {
    font-family: "OpentypeJsPreview";
    src: url('${URL.createObjectURL(new Blob([font.toArrayBuffer()]))}');
}
`;
```

For a font editor, you could have a live preview, not having to rely on the not-so-complete rendering capabilities of our library, actually pretty cool. I don't think it would hurt to add a helper method for comfortability, but don't think we're in a hurry to implement that.

Therefore I'll close this issue now.

**Comment by Pomax** [comment]
@Connum That sure sounds like something that _needs documenting and surfacing_ =)

**Comment by ILOVEPIE** [comment]
It's pretty self-explanatory.

**Comment by Connum** [comment]
Maybe we're overestimating the JS-savvyness of our user base. I definitely want to add the functionality at some point, it's just not a priority right now. Let's keep the issue open then.

**Comment by Pomax** [comment]
I can at least confirm that the JS-savvyness ranges from "seasoned web typography pro with a day-one typedrawers account" to "literally someone's first foray into finally using JS for something" =)
